File: src/service_example/service_client_sync/src/service_client_sync.cc
/******************************************************************************
 * Copyright (c) 2022-2026 SEGAR. All Rights Reserved.
 * SPDX-License-Identifier: LicenseRef-Segar-Proprietary
 *
 * PROPRIETARY AND CONFIDENTIAL. See ./LICENSE
 * for license terms and restrictions.
 *****************************************************************************/

#include <memory>

#include "example/srv/SetCameraInfo.hpp"

#include "segar/segar.h"

int main(int argc, char* argv[]) {
  RETURN_VAL_IF(!rti::segar::Init(argv[0]), EXIT_FAILURE);

  auto node = rti::segar::CreateNode("set_camera_info_client_sync");
  RETURN_VAL_IF(!node, EXIT_FAILURE);

  using SetCameraInfo = example::srv::SetCameraInfo;
  auto client = node->CreateClient<SetCameraInfo>("set_camera_info");
  RETURN_VAL_IF(!client, EXIT_FAILURE);

  uint32_t index = 0;
  auto callback = [&client, &index]() {
    auto request = std::make_shared<SetCameraInfo::Request>();
    request->camera_info().width(index);

    auto response = client->SyncSendRequest(request);
    if (response != nullptr) {
      AINFO << "[Sync] request camera width: " << index
            << ", Response msg:" << response->status_message();
    } else {
      AINFO << "[Sync] Request failed, request camera width: " << index;
    }

    index++;
  };

  // 1hz
  auto timer = std::make_shared<rti::segar::Timer>(1000, callback, false);
  timer->Start();
  rti::segar::WaitForShutdown();
  return EXIT_SUCCESS;
}

File: src_python/service_example/service_client_sync/src/service_client_sync.py
#!/usr/bin/env python3

import time

from example.msg import CameraInfo
from example.srv import SetCameraInfo
from segar.python import segar


def main():
    if not segar.Init("set_camera_info_client_sync"):
        return 1
    node = segar.CreateNode("set_camera_info_client_sync")
    if node is None:
        return 1

    client = node.CreateClient("set_camera_info", SetCameraInfo)
    if client is None:
        print("[service_client_sync] create client failed")
        return 1

    index = 0
    while segar.OK():
        request = SetCameraInfo.Request(camera_info=CameraInfo(width=index))
        response = client.SyncSendRequest(request)
        if response is not None:
            print(
                f"[service_client_sync] request width={index}, "
                f"response={response.status_message}"
            )
        else:
            print(f"[service_client_sync] request failed, width={index}")
        index += 1
        time.sleep(1.0)

    segar.WaitForShutdown()
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
